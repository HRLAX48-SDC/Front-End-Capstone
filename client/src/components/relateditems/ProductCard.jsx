import React, { useState, useEffect } from 'react';
import ActionButton from './ActionButton.jsx';
import axios from 'axios';
import API_KEY from '../../config/config.js';
import Card from 'react-bootstrap/Card';
import Button from 'react-bootstrap/Button';
import { Rating } from 'react-simple-star-rating';

class ProductCard extends React.Component {
  _isMounted = false;
  constructor(props) {
    super(props);

    this.state = {
      productid: this.props.product.id,
      category: this.props.product.category.toUpperCase(),
      imageUrl: '',
      // styleCall.results[0].photos[0].thumbnail_url;
      expandedProductName: this.props.product.name + ' - ' + this.props.product.slogan,
      price: this.props.product.default_price,
      rating: null
    }
  }

  componentDidMount(){
    this._isMounted = true;
    if(window.localStorage[`${this.props.product.id}/styles`]) {
      let currentStyle = JSON.parse(window.localStorage[`${this.props.product.id}/styles`]);
      this.setState({
        imageUrl: currentStyle.results[0].photos[0].thumbnail_url
      })
    } else {
      axios.get(`https://app-hrsei-api.herokuapp.com/api/fec2/hr-lax/products/${this.props.product.id}/styles`, {
      headers: {
        'Authorization': `${API_KEY}`
        }
      })
      .then((product) => {
        if(this._isMounted){
          if(product.data.results[0].photos[0].thumbnail_url){
            window.localStorage[`${this.props.product.id}/styles`] = JSON.stringify(product.data);
            this.setState({
              imageUrl: product.data.results[0].photos[0].thumbnail_url
            })
          } else {
            this.setState({
              imageUrl: `images/placeholder-image.png`
            })
          }
        }
      })
      .catch((err) => {
        console.log(err)
      })
    }
    if(window.localStorage[`meta/?product_id=${this.props.product.id}`]){
      let metaData = JSON.parse(window.localStorage[`meta/?product_id=${this.props.product.id}`]);
      let star1 = metaData.ratings[1] ? Number(metaData.ratings[1]) : 0;
      let star2 = metaData.ratings[2] ? Number(metaData.ratings[2]) : 0;
      let star3 = metaData.ratings[3] ? Number(metaData.ratings[3]) : 0;
      let star4 = metaData.ratings[4] ? Number(metaData.ratings[4]) : 0;
      let star5 = metaData.ratings[5] ? Number(metaData.ratings[5]) : 0;
      let total = star1 + star2 + star3 + star4 + star5;
      let rating = ((star1 * 1) + (star2 * 2) + (star3 * 3)+ (star4 * 4) + (star5 * 5)) / total;
      this.setState({
        rating: rating
      })
    }
    axios.get(`https://app-hrsei-api.herokuapp.com/api/fec2/hr-lax/reviews/meta/?product_id=${this.props.product.id}`, {
    headers: {
      'Authorization': `${API_KEY}`
      }
    })
    .then((meta) => {
      window.localStorage[`meta/?product_id=${this.props.product.id}`] = JSON.stringify(meta.data);
      let star1 = meta.data.ratings[1] ? Number(meta.data.ratings[1]) : 0;
      let star2 = meta.data.ratings[2] ? Number(meta.data.ratings[2]) : 0;
      let star3 = meta.data.ratings[3] ? Number(meta.data.ratings[3]) : 0;
      let star4 = meta.data.ratings[4] ? Number(meta.data.ratings[4]) : 0;
      let star5 = meta.data.ratings[5] ? Number(meta.data.ratings[5]) : 0;
      let total = star1 + star2 + star3 + star4 + star5
      let rating = ((star1 * 1) + (star2 * 2) + (star3 * 3)+ (star4 * 4) + (star5 * 5)) / total;
      if(this._isMounted){
        this.setState({
          rating: rating
        })
      }
    })
    .catch((err) => {
      console.log(err)
    })
  }

  componentWillUnmount(){
    this._isMounted = false;
  }

  render() {
    const { chooseProduct } = this.props;
    const { productid } = this.state;

    return (
      <Card style={{width: '18rem'}} onClick={() => {
        this.props.checkPos();
        this.props.chooseProduct(productid);
      }}>
        <Card.Img variant="top" src={this.state.imageUrl} style={styles.cardImage}/>
        <ActionButton icon={this.props.icon} removeOutfit={this.props.removeOutfit} product={this.props.product} featuredProd={this.props.featuredProd} />
        <Card.Body style={styles.cardText}>
          <Card.Text style={styles.noMargin}>{this.state.category}</Card.Text>
          <Card.Title>{this.state.expandedProductName}</Card.Title>
          <Card.Text style={styles.noMargin}>${this.state.price}</Card.Text>
          <Card.Text style={styles.noMargin}>
            <Rating readonly={true}
              ratingValue={this.state.rating * 20}
              initialValue={this.state.rating * 20}
              allowHalfIcon={true}
              size={20}
            />
          </Card.Text>
        </Card.Body>
      </Card>
    )
  }
}

export default ProductCard;

const styles = {
  cardImage: {
    maxWidth: '100%',
    height: '300px',
    overflow: 'hidden'
  },
  cardText: {
    whiteSpace: 'normal',
    textAlign: 'left'
  },
  noMargin: {
    marginBottom: 0
  }
}