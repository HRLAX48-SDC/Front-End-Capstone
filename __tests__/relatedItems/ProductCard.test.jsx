import React from 'react';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import TestRenderer from 'react-test-renderer';
import ProductCard from '../../client/src/components/relateditems/ProductCard.jsx';

Enzyme.configure({ adapter: new Adapter() });

let product;
let featuredProd;
let card;
const checkPosMock= jest.fn();
const chooseProductMock = jest.fn();
const resetPosMock = jest.fn();

beforeEach(async () => {
  product = {
    id: 42366,
    campus: "hr-lax",
    name: "Camo Onesie",
    slogan: "Blend in to your crowd",
    description: "The So Fatigues will wake you up and fit you in. This high energy camo will have you blending in to even the wildest surroundings.",
    category: "Jackets",
    default_price: "140.00",
    created_at: "2021-08-13T14:39:39.968Z",
    updated_at: "2021-08-13T14:39:39.968Z",
    features: [
      {
        feature: "Fabric",
        value: "Canvas"
      },
      {
        feature: "Buttons",
        value: "Brass"
      }
    ]
  };

  featuredProd = {
    id: 42367,
    campus: "hr-lax",
    name: "Bright Future Sunglasses",
    slogan: "You've got to wear shades",
    description: "Where you're going you might not need roads, but you definitely need some shades. Give those baby blues a rest and let the future shine bright on these timeless lenses.",
    category: "Accessories",
    default_price: "69.00",
    created_at: "2021-08-13T14:39:39.968Z",
    updated_at: "2021-08-13T14:39:39.968Z",
    features: [
      {
        "feature": "Lenses",
        "value": "Ultrasheen"
      },
      {
        "feature": "UV Protection",
        "value": null
      },
      {
        "feature": "Frames",
        "value": "LightCompose"
      }
    ]
  };

  card = mount(<ProductCard
    product={product}
    featuredProd={featuredProd}
    icon='remove'
    checkPos={checkPosMock}
    removeOutfit={() => {}}
    chooseProduct={chooseProductMock}
    resetPos={resetPosMock}
    uniqueid='42366'
  />);
})


test('renders Product Card component correctly', () => {
  const component = card.find('.card');

  expect(component).toExist();
})

test('click works', () => {
  const component = card.find('.card');
  card.simulate('click');

  expect(checkPosMock.mock.calls.length).toBe(1);
  expect(chooseProductMock.mock.calls.length).toBe(1);
})