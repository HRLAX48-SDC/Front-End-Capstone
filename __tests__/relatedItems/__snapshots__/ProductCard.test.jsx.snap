// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Card renders Product Card component correctly 1`] = `
<div
  className="card"
  onClick={[Function]}
  style={
    Object {
      "width": "18rem",
    }
  }
>
  <img
    className="card-img-top"
    src=""
    style={
      Object {
        "height": "300px",
        "maxWidth": "100%",
        "overflow": "hidden",
      }
    }
  />
  <div
    onClick={[Function]}
  >
    <span
      className="actionButton"
      onClick={[Function]}
    >
      <i
        className="far fa-times-circle"
      />
    </span>
  </div>
  <div
    className="card-body"
    style={
      Object {
        "textAlign": "left",
        "whiteSpace": "normal",
      }
    }
  >
    <p
      className="card-text"
      style={
        Object {
          "marginBottom": 0,
        }
      }
    >
      ACCESSORIES
    </p>
    <div
      className="card-title h5"
    >
      Bright Future Sunglasses - You've got to wear shades
    </div>
    <p
      className="card-text"
      style={
        Object {
          "marginBottom": 0,
        }
      }
    >
      $
      69.00
    </p>
    <p
      className="card-text"
      style={
        Object {
          "marginBottom": 0,
        }
      }
    >
      <span
        style={
          Object {
            "direction": "ltr",
            "display": "inline-block",
            "touchAction": "none",
          }
        }
      >
        <span
          aria-hidden="true"
          className="react-simple-star-rating"
          style={
            Object {
              "cursor": "",
              "display": "inline-block",
              "overflow": "hidden",
              "position": "relative",
              "userSelect": "none",
              "verticalAlign": "middle",
              "whiteSpace": "nowrap",
            }
          }
        >
          <span
            className="empty-icons"
            style={
              Object {
                "color": "#cccccc",
                "display": "inline-block",
              }
            }
          >
            <svg
              className="star-svg"
              fill="currentColor"
              height={20}
              style={Object {}}
              viewBox="0 0 24 24"
              width={20}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={20}
              style={Object {}}
              viewBox="0 0 24 24"
              width={20}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={20}
              style={Object {}}
              viewBox="0 0 24 24"
              width={20}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={20}
              style={Object {}}
              viewBox="0 0 24 24"
              width={20}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={20}
              style={Object {}}
              viewBox="0 0 24 24"
              width={20}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
          </span>
          <span
            className="filled-icons"
            style={
              Object {
                "color": "#ffbc0b",
                "display": "inline-block",
                "left": 0,
                "overflow": "hidden",
                "position": "absolute",
                "top": 0,
                "transition": "",
                "whiteSpace": "nowrap",
                "width": "0%",
              }
            }
            title="0 out of 5"
          >
            <svg
              className="star-svg"
              fill="currentColor"
              height={20}
              style={Object {}}
              viewBox="0 0 24 24"
              width={20}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={20}
              style={Object {}}
              viewBox="0 0 24 24"
              width={20}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={20}
              style={Object {}}
              viewBox="0 0 24 24"
              width={20}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={20}
              style={Object {}}
              viewBox="0 0 24 24"
              width={20}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={20}
              style={Object {}}
              viewBox="0 0 24 24"
              width={20}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
          </span>
        </span>
      </span>
    </p>
  </div>
</div>
`;

exports[`Should correctly render the reviewBody 1`] = `undefined`;

exports[`renders Product Card component correctly 1`] = `
<div
  className="card"
  onClick={[Function]}
  style={
    Object {
      "width": "18rem",
    }
  }
>
  <img
    className="card-img-top"
    src=""
    style={
      Object {
        "height": "300px",
        "maxWidth": "100%",
        "overflow": "hidden",
      }
    }
  />
  <div
    onClick={[Function]}
  >
    <span
      className="actionButton"
      onClick={[Function]}
    >
      <i
        className="far fa-times-circle"
      />
    </span>
  </div>
  <div
    className="card-body"
    style={
      Object {
        "textAlign": "left",
        "whiteSpace": "normal",
      }
    }
  >
    <p
      className="card-text"
      style={
        Object {
          "marginBottom": 0,
        }
      }
    >
      ACCESSORIES
    </p>
    <div
      className="card-title h5"
    >
      Bright Future Sunglasses - You've got to wear shades
    </div>
    <p
      className="card-text"
      style={
        Object {
          "marginBottom": 0,
        }
      }
    >
      $
      69.00
    </p>
    <p
      className="card-text"
      style={
        Object {
          "marginBottom": 0,
        }
      }
    >
      <span
        style={
          Object {
            "direction": "ltr",
            "display": "inline-block",
            "touchAction": "none",
          }
        }
      >
        <span
          aria-hidden="true"
          className="react-simple-star-rating"
          style={
            Object {
              "cursor": "",
              "display": "inline-block",
              "overflow": "hidden",
              "position": "relative",
              "userSelect": "none",
              "verticalAlign": "middle",
              "whiteSpace": "nowrap",
            }
          }
        >
          <span
            className="empty-icons"
            style={
              Object {
                "color": "#cccccc",
                "display": "inline-block",
              }
            }
          >
            <svg
              className="star-svg"
              fill="currentColor"
              height={20}
              style={Object {}}
              viewBox="0 0 24 24"
              width={20}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={20}
              style={Object {}}
              viewBox="0 0 24 24"
              width={20}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={20}
              style={Object {}}
              viewBox="0 0 24 24"
              width={20}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={20}
              style={Object {}}
              viewBox="0 0 24 24"
              width={20}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={20}
              style={Object {}}
              viewBox="0 0 24 24"
              width={20}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
          </span>
          <span
            className="filled-icons"
            style={
              Object {
                "color": "#ffbc0b",
                "display": "inline-block",
                "left": 0,
                "overflow": "hidden",
                "position": "absolute",
                "top": 0,
                "transition": "",
                "whiteSpace": "nowrap",
                "width": "0%",
              }
            }
            title="0 out of 5"
          >
            <svg
              className="star-svg"
              fill="currentColor"
              height={20}
              style={Object {}}
              viewBox="0 0 24 24"
              width={20}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={20}
              style={Object {}}
              viewBox="0 0 24 24"
              width={20}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={20}
              style={Object {}}
              viewBox="0 0 24 24"
              width={20}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={20}
              style={Object {}}
              viewBox="0 0 24 24"
              width={20}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
            <svg
              className="star-svg"
              fill="currentColor"
              height={20}
              style={Object {}}
              viewBox="0 0 24 24"
              width={20}
            >
              <path
                d="M12,17.27L18.18,21l-1.64-7.03L22,9.24l-7.19-0.61L12,2L9.19,8.63L2,9.24l5.46,4.73L5.82,21L12,17.27z"
                fill="currentColor"
                stroke="none"
                strokeMiterlimit="10"
                strokeWidth={0}
              />
            </svg>
          </span>
        </span>
      </span>
    </p>
  </div>
</div>
`;
